rust/kernel/hw_random.rs

```diff
@@ -0,0 +1,200 @@
+};
+
+use core::{cell::UnsafeCell, marker::PhantomData, pin::Pin};
+
+/// This trait is implemented in order to provide callbacks to `struct hwrng`.
+pub trait HwrngOperations: Sized + 'static {
+    /// The methods to use to populate [`struct hwrng`].
+    const TO_USE: ToUse;
+
+    /// The pointer type that will be used to hold user-defined data type.
+    type Data: PointerWrapper = Box<Self>;
+
+    /// Initialization callback, can be left undefined.
+    fn init(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) -> Result {
+        Err(Error::EINVAL)
+    }
+
+    /// Cleanup callback, can be left undefined.
+    fn cleanup(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) {}
+
+    /// Read data into the provided buffer.
+    /// Drivers can fill up to max bytes of data into the buffer.
+    /// The buffer is aligned for any type and max is a multiple of 4 and >= 32 bytes.
+    fn read(
+        data: <Self::Data as PointerWrapper>::Borrowed<'_>,
+        buffer: &mut [i8],
+        wait: bool,
+    ) -> Result<i32>;
+}
+
+/// Registration structure for Hardware Random Number Generator driver.
+pub struct Registration<T: HwrngOperations> {
+    hwrng: UnsafeCell<bindings::hwrng>,
+    _p: PhantomData<T>,
+}
+
+impl<'a, T: HwrngOperations> Registration<T> {
+    fn init_hwrng(
+        hwrng: &mut bindings::hwrng,
+        name: &'a CStr,
+        quality: u16,
+        data: *const core::ffi::c_void,
+    ) {
+        hwrng.name = name.as_char_ptr();
+
+        hwrng.init = if T::TO_USE.init {
+            Some(init_callback::<T>)
+        } else {
+            None
+        };
+        hwrng.cleanup = if T::TO_USE.cleanup {
+            Some(cleanup_callback::<T>)
+        } else {
+            None
+        };
+        hwrng.data_present = None;
+        hwrng.data_read = None;
+        hwrng.read = Some(read_callback::<T>);
+
+        hwrng.priv_ = data as _;
+        hwrng.quality = quality;
+
+        // SAFETY: All fields are properly initialized as
+        // remaining fields `list`, `ref` and `cleanup_done` are already
+        // zeroed by `bindings::hwrng::default()` call.
+    }
+
+    /// Registers a hwrng device within the rest of the kernel.
+    ///
+    /// It must be pinned because the memory block that represents the registration.
+    fn register(&self) -> Result {
+        // SAFETY: register function requires initialized `bindings::hwrng`.
+        // It is called in `Registration<T>::new_pinned` after initialization of the structure
+        // that quarantees safety.
+        let ret = unsafe { bindings::hwrng_register(&mut *self.hwrng.get()) };
+
+        if ret < 0 {
+            return Err(Error::from_kernel_errno(ret));
+        }
+        Ok(())
+    }
+
+    /// Returns a registered and pinned, heap-allocated representation of the registration.
+    pub fn new_pinned(name: &'a CStr, quality: u16, data: T::Data) -> Result<Pin<Box<Self>>> {
```

Comment:
Is this the only public way to create a registration?

I think we need to support a public `new` followed by a public `register` so that people can embed these as fields in a pinned outer structure. (So that a single allocation is needed for the  outer object).

Reply:
Sure, I will add required ones.
Thanks.